### DISCUSSION TITLE
Titlegame upgrades

### DISCUSSION THREAD

**Discussion by EratoNysiad** [OpenTTD/OpenTTD#8884] [Features]
Currently, the titlegame gets replaced every year [via a competition](https://github.com/OpenTTD/title-screen-competition-1.11/discussions/3). I think it definitely could be better than this.

I presume we're all aware of [Rollercoaster Tycoon 2](https://www.youtube.com/watch?v=mEgET98jZks), in which the title screen scrolls through different viewports on the same game. It would be cool if OpenTTD could do something similar.

Another thing I'd love to see is if we could load NewGRFs in the title screen. Combining the two might result in a title screen looking like this: https://www.youtube.com/watch?v=EoGcAr5PyZI

I understand that NewGRFs are to be downloaded by the player, and may be updated or otherwise be made incompatible, so a title game with NewGRFs would obviously not be packaged with the main game, but a game with multiple different viewports definitely could. A titlegame with NewGRFs would have to be distributed on the forums, for example.

I'd love to hear what you lot think.

**Comment by EratoNysiad** [comment]
The different viewpoints could be arranged in a few different ways. I could really only think of 2:
1) A csv file with "x, y, duration"
2) Signs in the game that don't render, that you have to name "VIEWPORT 1: 12" for 12s, for instance

**Comment by James103** [comment]
OpenTTD has the ability to follow any vehicle as it travels when in main gameplay. OpenRCT2 has the ability to follow a sprite on the title screen. Why not allow OpenTTD title sequences to have segments where the viewport follows a given vehicle?

**Comment by James103** [comment]
Would it be possible to allow an OpenTTD title sequence to load a new map at some point in the title sequence? For example, one scene may be in a temperate map following a steam locomotive, then the next scene is in a tropical map showing off a Ro-Ro station transferring cargo from a water supply.

**Comment by LordAro** [comment]
A long time ago I made a (very hacky) patch to enable moving the titlegame viewport between different signs. It could certainly be improved and extended using suggestions above

https://github.com/OpenTTD/OpenTTD/compare/master...LordAro:movable-menuviewport

As nice as it would be, I would suggest that loading multiple titlegames is likely too much effort and infeasible to do well

**Comment by nielsmh** [comment]
One thing that might be possible is split the screen into multiple viewports and show e.g. 4 different positions at once, perhaps even have them follow different vehicles.

**Comment by nielsmh** [comment]
Okay some concrete ideas for how a "sign scripting language" for dynamic title screens could work.

The basic idea is that you have signs with commands that order how the viewport moves around. If there are no signs, then the existing behaviour of just using the saved position is used.

- All the signs controlling this start with a `T` followed by a space and an order number. The signs are deleted after the title save is loaded as title save, so they won't show up on screen.
- Because the signs are numbered, sometimes the title game designer might need to insert new signs in the middle, and that can be a hassle. The patch should provide a console command to renumber existing signs. (The same way as you had tools to renumber BASIC programs on old home computers.)
- The signs also need to have a delay, how long to keep the viewport there.
- There should be a way to control how the viewport is anchored to a sign: Top, middle, bottom, left, center, right. You can combine these in any way. Shorthand exclude one of the dimensions and the other is assumed middle/center.
- There should be a way to control zoom level. This should be limited to +1/0/-1, relative to the title screen base zoom level (controlled by the interface size option.)
- There should be a way to create panning effects.
- There should be a way to follow a vehicle.

Examples:
`T 1 C 10` - Order 1, center for 10 seconds.
`T 2 TC 10` - Order 2, align top-center for 10 seconds
`T 3 C 5 Z-` - Order 3, center for 5 seconds, zoomed out one step
`T 4 R 5 Z+` - Order 4, align middle-right for 5 seconds, zoomed in one step (relative to base zoom)
`T 5 CP 15` - Order 5, center at sign and pan towards the next order over 15 seconds (zoom is back to base, since none was specified)
`T 8 TL 0` - Order 8 (numbers 6 and 7 didn't exist, missing numbers are fine), align top-left for zero seconds (this is just an end-point anchor for the previous pan order)
`T 9 C 20 Z+ F "Train 3"` - Order 9, center on vehicle "Train 3", zoomed in, and follow it for 20 seconds

Ideas I'm not yet sure how to fit in: Multiple viewports. Whether maybe anchoring others than center on vehicles should work, and if they should maybe not use screen absolute but more like 25% and 75% positions for left/top and right/bottom positions. Following vehicles would probably be most interesting with multiple viewports.

**Reply by LordAro** [reply]
Very good ideas in general, but some follow ups:

* No need to actually delete the signs, just hide them like all other things on the titlegame - https://github.com/OpenTTD/OpenTTD/blob/3ede756d23b09f52a8099736d9a78abc966448b8/src/viewport.cpp#L1342-L1346
* Signs are numbered... but no need for a command to reorder them? Just encourage people to number them 10, 20, 30... Should be sufficient
* Is there any advantage to anchoring a viewport anywhere other than centred to the sign? I'm not sure I'd expect any other behaviour (except perhaps top-left like it is currently)

**Reply by nielsmh** [reply]
Anchoring: Since the main menu covers the center of the screen, anchoring to a side or corner gives some better control of what gets shown/focused, and might give better results when you have anchors near the edge of the map.

**Reply by nielsmh** [reply]
Oh, and the reason for deleting the signs is actually to allow the designer to put other signs that *will* show. I'm not sure if that will have value, but I thought it should be an option.

**Reply by nielsmh** [reply]
Posted PR #8980

**Comment by LC-Zorg** [comment]
The possibility that is most lacking, at least from my point of view, is the use of NewGRF in such a save. At least for personal use, but also why not for general use. :) Opntitle contests would be possible, same as Screenshot of the month. Or maybe sometimes it would be possible to use the save from the winner of the current contest if the author also makes the save available? :) It would be really interesting to be able to use these great works.

Here the question remains how to replace the current opntitle. Currently, it is a bit troublesome, especially for less knowledgeable players. There is also a risk that someone will break something during the change. It's nice if there was an option to choose a save in the game. For example, in the Game Options, clicking on the bar would open the game loading window and you could choose any save.

When it comes to changing views, it's a nice idea, but I remember from Springpatch 2013 that these changes were quite irritating over time. The problem is the sudden transition. For many other games, if different scenes are presented, the transitions between them are smooth. I don't know if it would be easy or hard (rather hard?) to add this, but in my opinion it would be very advisable.
